### PR TITLE
Implement alert for unsaved changes

### PR DESCRIPTION
**Pull request by jackyzy823**
For #296 

There's two model: 1) Tracking if changed (not very accurately ie: Enable and disable and then enable is still treated as changed) 2) Comparison between current and original

It looks like SmartProxy has implemented some basic objection for tracking (settingsPage.changeTracking). So I use the tracking model for most changes.

However , it's hard/(too complex) to apply tacking model on `GeneralOptions`  ,  But comparison is suitable here.

---------------

|Tracking| Action | Element  | Status |
|---------|-------|----------|--------|
| activeProxy | Change Profile | cmbActiveProxyServer | DONE |
|||
|smartProfiles| Edit profilename | txtSmartProfileName | DONE |
|smartProfiles| Toggle Enable/Disable | chkSmartProfileEnabled | DONE |
|smartProfiles| Choose Profile's Proxy Server | cmbProfileProxyServer | DONE |
|smartProfiles| Add/Edit rule | btnSubmitRule | DONE |
|smartProfiles| Add multiple rules | btnSubmitMultipleRule | DONE |
|smartProfiles| Import rule | btnImportRules | DONE |
|smartProfiles| Remove rule | btnRulesRemove | DONE |
|smartProfiles| Clear all rules | btnClearProxyRules | DONE |
|smartProfiles| Delete multiple rules | btnRemoveMultipleProxyRule  | DONE in #341 |
||||
|rulesSubscriptions| Subscribe/Edit a rules list | btnSaveRulesSubscriptions | Already |
|rulesSubscriptions| Delete rules list | btnRuleSubscriptionsRemove | Already |
|rulesSubscriptions| Clear all rules lists | btnClearRulesSubscriptions | Already |
|rulesSubscriptions| Refresh rules list | btnRuleSubscriptionsRefresh | Already |
|rulesSubscriptions| Delete multiple rules | btnRemoveMultipleRulesSubscription | DONE in #341 |
|||||
|servers| Add/Edit server | btnSubmitProxyServer | Already |
|servers| Delete server | btnServersRemove | Already |
|servers| Clear all servers | btnClearProxyServers | Already |
|servers| Import server | btnImportProxyServer | DONE |
|servers| Delete multiple servers | btnRemoveMultipleProxyServer |  DONE in #341 |
|||||
|serverSubscriptions| Subscribe list | btnSaveServerSubscription | Already |
|serverSubscriptions| Delete list | btnSubscriptionsRemove | Already |
|serverSubscriptions| Clear all lists | btnClearServerSubscriptions | Already |
|serverSubscriptions| Delete multiple lists | btnRemoveMultipleServerSubscription | DONE in #341 |
|||||
|SPECIAL| Import server backup | btnImportProxyServer | SAME as Import server |
|SPECIAL | Restore backup | btnRestoreBackup | IGNORE unsaved changes by removing eventListenr |
|SPECIAL | Factory Reset | btnFactoryReset | IGNORE unsaved changes by removing eventListenr |
|||
|SPECIAL| Delete a profile | btnDeleteSmartProfile | NOTHING (because runtime message is send directly) |
|SPECIAL | Add a new profile | | if user choose to save when staying , try save it (intent to fail if profile title is empty) |

Also `changeTracking` to false when runtimeSendMessage(SettingsPageSave*) is called.

Note: 
changeTracking .rulesSubscriptions should also set to false in onClickSaveSmartProfile  ( same as changeTracking.smartProfile )

### CODE REVIEW

**Comment by salarcode** [comment]
Thanks for this, I will check it soon.
I wish if you were to work on a feature request you worked on something higher priority, like these:
#273 , #317 , #342 and etc
If you want to work on anything please let me know I'll refine the issue details

**Comment by jackyzy823** [comment]
I may work on #273 , but implementation details should be discussed. (like export in what format or etc...)
 
I really don't understand  what #317 is for.
I won't work on #342 , because i don't have any mac devices.

Currently i'm working on `SwitchyOmega` import related stuff like #318.

**Comment by salarcode** [comment]
Thanks. I have already started working on #318 , #286 so that one is not needed

File: src/core/definitions.ts [hunk 1]
```diff
@@ -595,12 +595,37 @@ export class GeneralOptions implements Cloneable {
 		this.themesDark = source['themesDark'];
 		this.themesDarkCustomUrl = source['themesDarkCustomUrl'];
 	}
+
+	Equals(other: GeneralOptions): Boolean {
```
Comment (salarcode) [inline src/core/definitions.ts:599]: This method is not gonna work well. Settings could come from different locations like sync, backups and etc and might miss some properties.
This method should treat null/undefined and empty values as same

File: src/ui/code/settingsPage.ts [hunk 2]
```diff
@@ -2186,6 +2219,10 @@ export class settingsPage {
 			settingsPage.updateProfileGridsLayout(pageSmartProfile);
 			settingsPage.selectAddNewProfileMenu();
 
+			settingsPage.unsavedProfile = pageSmartProfile;
```
Comment (salarcode) [inline src/ui/code/settingsPage.ts:2222]: I'm not happy to keep a reference of `pageSmartProfile` in the page level, we can be more smart about this.

File: src/ui/code/settingsPage.ts [hunk 3]
```diff
@@ -3672,6 +3723,46 @@ export class settingsPage {
...
+							}
+							if (settingsPage.changeTracking.serverSubscriptions) {
+								settingsPage.uiEvents.onClickSaveServerSubscriptionsChanges();
+							}
+						});
+				}, 200);
+			});
+
+			// Browser will show a dialog asking user to leave or stay.
+			event.preventDefault();
```
Comment (salarcode) [inline src/ui/code/settingsPage.ts:3764]: Is there a way to customize the confirm message that browsers are showing? I did not see any!

**Review by salarcode — CHANGES_REQUESTED** [review]

**Comment by salarcode** [comment]
Thanks for the write up, overall looking good but I've left some comments

File: src/ui/code/settingsPage.ts (hunk 2 above)
Comment (jackyzy823) [reply src/ui/code/settingsPage.ts:2222]: Yes, i don't like this too, but i couldn't  find a way to find if there's  an unsaved profile (and get its page object)  when leaving the page.

File: src/core/definitions.ts (hunk 1 above)
Comment (salarcode) [reply src/core/definitions.ts:599]: with `neq` function it should now

**Pull request merged**